File: Linked_List/linked.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
        self.prev = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def add_to_front(self, value):
        new_node = Node(value)
        new_node.next = self.head
        if self.head:
            self.head.prev = new_node
        else:
            self.tail = new_node
        self.head = new_node

    def add_to_end(self, value):
        new_node = Node(value)
        new_node.prev = self.tail
        if self.tail:
            self.tail.next = new_node
        else:
            self.head = new_node
        self.tail = new_node

    def remove_from_front(self):
        if not self.head:
            return None

        removed_value = self.head.value
        self.head = self.head.next

        if self.head:
            self.head.prev = None
        else:
            self.tail = None
        return removed_value

    def remove_from_end(self):
        if not self.tail:
            return None

        removed_value = self.tail.value
        self.tail = self.tail.prev

        if self.head:
            self.head.prev = None
        else:
            self.tail = None

        return removed_value


linked_list = DoublyLinkedList()
linked_list.add_to_front(10)
linked_list.add_to_front(25)
linked_list.add_to_front(30)

linked_list.add_to_end(5)
linked_list.add_to_end(2)

print(linked_list.remove_from_front())
print(linked_list.remove_from_end())

File: Linked_List/linked.java
class Node {
    int value;
    Node next;
    Node prev;

    public Node(int value) {
        this.value = value;
        this.next = null;
        this.prev = null;
    }
}

class DoublyLinkedList {
    private Node head;
    private Node tail;

    public DoublyLinkedList() {
        this.head = null;
        this.tail = null;
    }

    public void addToFront(int value) {
        Node newNode = new Node(value);
        newNode.next = head;
        if (head != null) {
            head.prev = newNode;
        } else {
            tail = newNode;
        }
        head = newNode;
    }

    public void addToEnd(int value) {
        Node newNode = new Node(value);
        newNode.prev = tail;
        if (tail != null) {
            tail.next = newNode;
        } else {
            head = newNode;
        }
        tail = newNode;
    }

    public Integer removeFromFront() {
        if (head == null) {
            return null;
        }

        int removedValue = head.value;
        head = head.next;

        if (head != null) {
            head.prev = null;
        } else {
            tail = null;
        }

        return removedValue;
    }

    public Integer removeFromEnd() {
        if (tail == null) {
            return null;
        }

        int removedValue = tail.value;
        tail = tail.prev;

        if (tail != null) {
            tail.next = null;
        } else {
            head = null;
        }

        return removedValue;
    }
}

public class Main {
    public static void main(String[] args) {
        DoublyLinkedList linkedList = new DoublyLinkedList();

        linkedList.addToFront(10);
        linkedList.addToFront(25);
        linkedList.addToFront(30);

        linkedList.addToEnd(5);
        linkedList.addToEnd(2);

        System.out.println(linkedList.removeFromFront()); // Output: 30
        System.out.println(linkedList.removeFromEnd());   // Output: 2
    }
}
